### PR TITLE
[BAHIR-72][bahir-flink] support netty: pushed tcp/http connector

### PR DESCRIPTION
**Pull request by shijinkui**
When source stream get start, listen a provided tcp port, receive stream data from user data source.
This netty tcp source is keepping alive and end-to-end, that is from business system to flink worker directly.
1.  source run as a netty tcp and http server
2.  user provide a tcp port, if the port is in used, increace the port number between 1024 to 65535. Source can parallel.
3.  callback the provided url to report the real port to listen
4.  user push streaming data to netty server, then collect the data to flink

### CODE REVIEW

File: flink-connector-netty/pom.xml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-all</artifactId>
+      <version>4.1.5.Final</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.flink</groupId>
+      <artifactId>flink-table_${scala.binary.version}</artifactId>
+      <version>${flink.version}</version>
+    </dependency>
```
Comment (rmetzger) [inline flink-connector-netty/pom.xml]: I don't think its a good idea to add the Table API here as a dependency just for one connector example.

Flink Table could grow quite big in the future, and users of the connector will not execute a single line of the table API.

File: flink-connector-netty/src/main/scala/org/apache/flink/streaming/connectors/netty/example/NettyUtil.scala [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        val result = startService(tryPort)
+        logger.info(s"Successfully started service$serviceString, result:$result.")
+        return result
+      } catch {
+        case e: Exception if isBindCollision(e) =>
+          if (offset >= maxRetries) {
+            val exceptionMessage = s"${e.getMessage}: Service$serviceString failed after " +
+              s"$maxRetries retries! Consider explicitly setting the appropriate port for the " +
+              s"service$serviceString (for example spark.ui.port for SparkUI) to an available " +
+              "port or increasing spark.port.maxRetries."
```
Comment (rmetzger) [inline flink-connector-netty/src/main/scala/org/apache/flink/streaming/connectors/netty/example/NettyUtil.scala]: The exception message doesn't seem to be relevant to Flink.

In general, it seems that the code here seems to be copied from Apache Spark: https://github.com/apache/spark/blob/39755169fb5bb07332eef263b4c18ede1528812d/core/src/main/scala/org/apache/spark/util/Utils.scala#L2172

Can you add comments to the code copied from other projects?

Comment (shijinkui) [inline flink-connector-netty/src/main/scala/org/apache/flink/streaming/connectors/netty/example/NettyUtil.scala]: @rmetzger OK, added

Comment (shijinkui) [inline flink-connector-netty/src/main/scala/org/apache/flink/streaming/connectors/netty/example/NettyUtil.scala]: @rmetzger I've added example to TcpReceiverSource and HttpReceiverSource. Also there are two full example in test package `TcpSourceExample` and `StreamSqlExample`

**Comment by shijinkui** [comment]
@rmetzger any problem?

**Comment by rmetzger** [comment]
Thanks a lot for the updates. I was traveling the last few weeks, that's why I didn't have time to review your changes earlier.

I'll merge the PR.

**Comment by ckadner** [comment]
Sadly we did not have *Scalatest* enabled at the time this PR was reviewed, so we missed adding automated unit tests.

I opened [BAHIR-113: Flink Netty connector missing automated unit tests](https://issues.apache.org/jira/browse/BAHIR-113) to keep track of that.

@shijinkui -- would you be willing to take that on and open a PR for  [BAHIR-113](https://issues.apache.org/jira/browse/BAHIR-113)?
